zbus/src/connection/mod.rs

```diff
@@ -1217,6 +1229,24 @@ impl Connection {
             .map_err(Into::into)
     }
 
+    /// Gracefully close the connection, waiting for all other references to be dropped.
+    ///
+    /// This will not disrupt any incoming or outgoing method calls, and will await their
+    /// completion.
+    ///
+    /// # Caution
+    ///
+    /// * Calling this will not prevent new incoming calls from arriving and keeping the connection
+    ///   alive, so if this is undesirable you may want to deregister any well-known names first.
```

Comment:
hmm.. I think this is a problem that we can solve ourselves in zbus so let's do that as part of this method and document that we do this. The names will be lost anyway once this method is done, so it's very much a desired effect and something caller will want in all cases.

Reply:
I have been thinking about this. Some thoughts: 

- I'm not sure this is really an issue worth mentioning even. The user is explicitly asking for shutdown **after** all pending methods are handled and they can wait for that. Except in case of some spamming client it long running method handles, this is unlikely to take long in most cases. 
- Well-known names aren't the only way for clients to make method calls into the service, so deregistering names isn't an effective way to stop further incoming calls.
- if we really want to effectively stop further method calls, we should find a way to cancel the ObjectServer task instead.

My suggestion would be to not bother with this for now. If this actually becomes an issue, we can solve it then. For now, just remove this paragraph.

Reply:
>> if we really want to effectively stop further method calls, we should find a way to cancel the ObjectServer task instead.
>
> This is an interesting idea. I guess we'd have to introduce a second event in ConnectionInner that signals the object server to stop listening.

Thinking about this more, I don't think such a feature would be useful. Presumably the dbus broker would be free to send us more method calls even if we're not listening to the socket anymore, and those method calls would get dropped ungracefully, at which point the whole thing is a bit pointless.

Are you happy with the rephrasing I did?

Reply:
> * If my understanding is correct, releasing a well-known name is in practice _almost_ a guarantee you're not going to receive any more method calls, since the only other way is if someone knew your arbitrarily-assigned unique name, right?

That's not 100% correct. Firstly, there is nothing stopping clients from caching the unique name of the service and calling their methods directly to the unique name. 

Secondly, we need to consider what will happen with and without this change. Clients will still make their calls (most services, even autostart when new messages come for them: a d-bus feature) regardless of whether or not the name in question is taken currently. Let's look at all scenarios here:

  1. If the service has **not** set itself up for autolaunch:
    1. if the service manages to successfully release its name,  the client will get an error from the bus.
    2. if the service does not manage to release its name before it can exit, the client will get an error from the service instead of the bus.
  2. If the server **has** set itself up for autolaunch:
    1. if the service manages to successfully release its name,  the bus will launch the service in a new process (which may not be desirable).
    2. if the service does not manage to release its name before it can exit, the bus will launch the service here too but in this case, there's only going to be one process of the service. I believe this is what people would expect to happen.

So I really don't think we should be recommending releasing of name before exiting.

Reply:
Just to add that typically services don't really care that much about messages not getting handled when they're shutting down, because of how D-Bus works (as I explained a bit above). I understand that in some cases, service authors would like to handle the existing queue of calls (and it doesn't hurt to do so in general) so the main point here is to add a best-effort graceful shutdown.

Reply:
> there is nothing stopping clients from caching the unique name of the service and calling their methods directly to the unique name.

True, although this would be brittle to service restarts resulting in a new unique name, so would any client actually do this?

> Let's look at all scenarios here...

I think there's one case missing from your analysis, which is your case 2.1 but with systemd as a service manager and a service of `Type=dbus` - in this case systemd will consider the service to be "stopping" as soon as the name is released. A subsequent call resulting in autolaunch will queue a systemd start behind the stop, bringing the service back up to handle the call as soon as it exits. This results in the stop-start being entirely transparent to the caller and the service effectively being available uninterrupted, and is precisely the example motivating me.

(I think there _might_ be a tiny race still, where if the method call comes in after the dbus broker thinks the name is gone but before systemd has processed the same fact, then the autolaunch could be absorbed as a no-op since systemd thinks the service is still running, resulting in an eventual timeout. But that window seems vanishingly small).

However, you've convinced me that this is niche enough not to be worth making as a general recommendation; I will remove the wording from the docs.

Reply:
> > there is nothing stopping clients from caching the unique name of the service and calling their methods directly to the unique name.
> 
> True, although this would be brittle to service restarts resulting in a new unique name, so would any client actually do this?

I agree that they really should not do this but this is the wild world of D-Bus. I hear of people doing all kinds of strange things. :)

> I think there's one case missing from your analysis, which is your case 2.1 but with systemd as a service manager and a service of `Type=dbus`

Right and I've to admit, I didn't remember how systemd handles this. However, this is not the problematic case so we don't need to worry about it IMO.

> (I think there _might_ be a tiny race still, where if the method call comes in after the dbus broker thinks the name is gone but before systemd has processed the same fact, then the autolaunch could be absorbed as a no-op since systemd thinks the service is still running, resulting in an eventual timeout. But that window seems vanishingly small).

True. While the window is small, we shouldn't rely on it being non-problematic. I think no matter what we and service impls do here, unfortunately shutdown can't be guaranteed to be 100% graceful.



> However, you've convinced me that this is niche enough not to be worth making as a general recommendation; I will remove the wording from the docs.

:+1: